crypto/benches/mls_proteus.rs

```diff
@@ -0,0 +1,268 @@
+    let in_memory = (ciphersuite, credential.clone(), true, "MLS/mem");
+    let in_db = (ciphersuite, credential, false, "MLS/db");
+    if cfg!(feature = "bench-in-db") {
+        vec![in_memory, in_db]
+    } else {
+        vec![in_memory]
+    }
+}
+
+fn proteus_cases() -> Vec<(bool, &'static str)> {
+    let in_memory = (true, "Proteus/mem");
+    let in_db = (false, "Proteus/db");
+    if cfg!(feature = "bench-in-db") {
+        vec![in_memory, in_db]
+    } else {
+        vec![in_memory]
+    }
+}
+
+fn encrypt_message_bench(c: &mut Criterion) {
+    let mut group = c.benchmark_group("Mls vs Proteus: encrypt");
+    for i in (GROUP_RANGE).step_by(GROUP_STEP) {
+        // MLS
+        for (ciphersuite, credential, in_memory, bench_name) in mls_cases() {
+            group.bench_with_input(BenchmarkId::new(bench_name, i), &i, |b, i| {
+                b.to_async(FuturesExecutor).iter_batched(
+                    || {
+                        let (mut central, id) = setup_mls(ciphersuite, &credential, in_memory);
+                        add_clients(&mut central, &id, ciphersuite, *i);
+                        let text = Alphanumeric.sample_string(&mut rand::thread_rng(), MSG_MAX);
+                        (central, id, text)
+                    },
+                    |(mut central, id, text)| async move {
+                        black_box(central.encrypt_message(&id, text).await.unwrap());
+                    },
+                    BatchSize::SmallInput,
+                )
+            });
+        }
+        // Proteus
+        for (in_memory, bench_name) in proteus_cases() {
+            group.bench_with_input(BenchmarkId::new(bench_name, i), &i, |b, i| {
+                b.to_async(FuturesExecutor).iter_batched(
+                    || {
+                        let (mut central, keystore) = setup_proteus(in_memory);
+                        let session_material = (0..*i)
+                            .map(|_| (session_id(), new_prekey().serialise().unwrap()))
+                            .collect::<Vec<(String, Vec<u8>)>>();
+                        for (session_id, key) in &session_material {
+                            block_on(async {
+                                central.session_from_prekey(session_id, key).await.unwrap();
+                                central.session_save(&keystore, session_id).await.unwrap();
+                            });
+                        }
+                        let text = Alphanumeric.sample_string(&mut rand::thread_rng(), MSG_MAX);
+                        (central, keystore, session_material, text)
+                    },
+                    |(mut central, keystore, session_material, text)| async move {
+                        for (session_id, _) in session_material {
+                            black_box(central.encrypt(&session_id, text.as_bytes()).unwrap());
+                            black_box(central.session_save(&keystore, &session_id).await.unwrap());
```

Comment:
Not sure if the session_save call is needed here as you're just encrypting payloads. Might lead to false I/O overhead.

Reply:
I think saving the session is required because `encrypt` requires a mutable session (to ratchet [here](https://github.com/wireapp/proteus/blob/develop/src/internal/session.rs#L864) I think).  
I wanted to test both in "safe" mode (we also persist the group in MLS after encrypting) to have a fair benchmark